ctk/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt

```diff
@@ -129,74 +135,78 @@ class RedirectBindingVerifier(private val response: IdpRedirectResponseDecorator
      * binding spec
      * 3.4.4 Message Encoding
      */
-    private fun verifyNoNulls() {
-        with(response) {
-            if (isUrlNull) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "Url not found.")
-            }
-            if (isPathNull) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "Path not found.")
-            }
-            if (isParametersNull) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "Parameters not found.")
-            }
-            if (samlResponse == null) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "SAMLResponse not found.")
-            }
-            if (isRelayStateGiven && relayState == null) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_3_b,
-                        message = "RelayState not found.")
-            }
+    private fun verifyNoNullsAndParse(): Map<String, String> {
+        val url = response.header(LOCATION) ?: throw SAMLComplianceException.create(
+                SAMLBindings_3_4_4_b,
+                message = "Url not found.")
+
+        val splitUrl = Splitter.on("?").splitToList(url)
+
+        splitUrl[0] ?: throw SAMLComplianceException.create(
+                SAMLBindings_3_4_4_b,
+                message = "Path not found.")
+
+        val parameters = splitUrl[1] ?: throw SAMLComplianceException.create(
+                SAMLBindings_3_4_4_b,
+                message = "Parameters not found.")
+
+        val paramMap = parameters.split("&")
+                .map { s -> s.split("=") }
+                .associate { s -> s[0] to s[1] }
```

Comment:
❓ This syntax seems odd to me. I want to make sure it isn't an error.

Assume we have a List, `list`, and elements in the list `0, 1, 2, 3, 4`, and we run `list.associate { s -> s[0] to s[1] }`.
It looks like this associate will execute the same statement (`s[0] to s[1]`) 5 times, resulting in a map of value `0 -> 1` and no other elements.

Reply:
No, in this scenario the input is a list of arrays. So a good example is `[0, 1], [2, 3], [4, 5]`. That turns into the key-value pairs `0 -> 1, 2 -> 3, 4 -> 5`.

The assumption we're making here is that we'll never have more than 2 elements in each of those arrays. Which is a fair assumption since it wouldn't be a valid URL at that point. It'd have to be something like `https://hostname:123/path?VALID=valid&INVALID=invalid=invalid=invalid`